### ISSUE TITLE
Entity context API

### ISSUE THREAD

**Issue by JDGrimes** [WordPoints/hooks-api#63]
Forked from https://github.com/WordPoints/hooks-api/issues/59#issuecomment-167364741:

> ## Context-specific entities
> 
> There is one more issue that needs to be considered when defining event repeatability: some entities are not global, but context-specific.
> 
> What is meant by the entity being "local" to a particular context is that entities of that type can exist within the scope of the reaction, such that two such entities with the same ID may exist in separate sub-contexts.
> 
> We discussed this above with network-wide reactions and posts. Posts are entities that exist in the context of a specific site, not the network as a whole. This means that each site may have a post with the ID 2, for example, but because these are different posts, each with its own status, even when an event is non-repeatable, the event isn't restricted to occurring only once per post ID, but once per post per site on the network.
> 
> Stated generally, a fire of an event is only considered a repeat if it has already occurred for an entity with the ID of the signature arg _in the context in which that entity exists_.
> 
> In other words, the repeat must be calculated relative to that particular entity, not just to its ID.
> 
> The question, then, is how we do should do this. Should we try to introduce a full-fledged "contexts" API? Or should we just allow the entity class to supply a totally unique ID that can be used across all contexts?
> 
> Or, are we really going to have other contexts beside the site context that we run into on multisite? If not, we could build the API specifically for that.

**Comment by JDGrimes** [comment]
I think it makes sense for this info to be supplied by the entity object. But just supplying a global UID won't allow us to determine what specific context that entity is in. It wouldn't allow us to determine whether two entities were in the same context, for example. I'm not sure that that is important at the moment, but it may become so in the future.

## Multisite only?

Building just for multisite doesn't seem wise, unless we can prove that there are no other possible contexts that we'd ever want to support. And I think that is a non-starter. Even if I can't think of one at the moment, I'd say there is just about bound to be other types of contexts. So going the generic route from the start seems wise.

## Nested contexts

There is also the possibility of nested contexts. For example, there may be several different contexts for an entity on each site. Then on multisite where each site in turn has a different context, we might say these are nested contexts.

But is it really important that the context is nested? Couldn't we just consider the most nested context when calculating the GUID? But then you could have conflicts between sub-contexts in different sites. So I guess it really depends upon the nature of the context. If it is truly a sub-context of the site, then it could have a twin on any site on the network. If it couldn't have a twin on another site, then it isn't really in the sub-context of a particular site, but of the network itself.

## Install GUIDs

A question that we need to consider is how to calculate the GUIDs of installs. What I mean by this is that we might need to be able to identify the difference between a post from the current site and one from an entirely separate WordPress installation. We might want our GUIDs to be truly global, not just local to the current install (whether a single or multisite/multi-network install). I don't think WordPress installs really have GUIDs themselves, which presents a problem.

But then, I suppose we'd need to have a different entity class for a post that was coming from another install, because we wouldn't be able to retrieve the data for it in the same way. What this means is that entity classes that represent entities which come from the current install can't currently be used to represent entities that come from other installs.

Another thing to consider is that we'd have to just trust whatever GUID was given us by the other install anyway. We wouldn't be able to verify it. What that basically means is that it would make it very easy for the user to cheat the repeat detection system by supplying made up GUIDs. 

In other words, GUIDs would only fully protect us from repeat fires within the context of the install; added event validation measures would need to be taken for any action that is triggered from the outside. Put another way, all external events are repeatable by default, since we don't provide the infrastructure to detect and stop repetitions.

More pertinent to the current ticket is that we don't currently have a way of calculating GUIDs that will be unique outside of the context of the current install. That may be something that we'll want to pursue in the future, but at the moment it isn't important enough to spend the time on.

**Comment by JDGrimes** [comment]
A note in #62 reminds me that we may want to consider somehow disabling entity operation outside of an entity's context. Related to that is #61.

**Comment by JDGrimes** [comment]
I've just [discovered](http://wptavern.com/automattics-wordads-network-is-now-open-to-self-hosted-wordpress-sites) that Jetpack assigns each site a unique ID. So it is doable, but likely requires that there be a central server which is "phoned home" to. We could investigate exactly how it works though, and maybe we'll find something useful.

**Comment by JDGrimes** [comment]
We might need to have the full context for all args in a relationship hierarchy for each signature arg for periods (#10) in case an entity can be related to several similar entities from different contexts.

**Comment by JDGrimes** [comment]
## Context data types

Since it is possible to have nested contexts, we'll likely need to represent the context as an array in many cases so that we can easily get the values of specific levels of the context. However, we'll also certainly need to get it in string form as well, for saving to the database, etc. This could be accomplished through serialization, but that might not be the best answer.

If we implode them into a string though, we will need to use a separator, which works well enough for the integer IDs of the sites and networks. But we might also have string slugs for the contexts, which might contain any sort of character, including our chosen separator.

So it might be best to just leave it as an array. The problem there comes when we need to search for something based on a specific level of the context, instead of the whole thing. But in that case I guess we'd want a completely different table structure than if we just stored the whole thing in a single row. (And I'm really not sure at this point whether we'd need such a feature or not.) However, it does seem to make sense that we let each application decide how it is best to store this info.

Remember also that each context has to be identified by a pair of things: the slug of that type of context, and the ID of that context. So an array of `slug => id` pairs makes sense. For example:

``` php
$context = array( 'network' => 1, 'site' => 5 );
```

The only issue there would be that when generating the GUID for an entity by combining the context and ID we'd break things when the context slug conflicted with an entity slug. Take `site`, for example. But then I suppose what we really have is a type of entity that also defines a context. So we would have `array( 'network' => 1, 'site' => 5, 'site' => 5 )` (which is impossible anyway), we'd really just want `array( 'network' => 1, 'site' => 5 )`.

**Comment by JDGrimes** [comment]
## Setting entity contexts

In my mind I've been thinking that we might want to make it possible to set the context of an entity, just as we set the value. https://github.com/WordPoints/hooks-api/issues/61#issuecomment-169135067 indicates that this may indeed be a useful feature. Without it, we have to assume that entities always come from the current context, which is not necessarily true (especially on multisite there may be practical applications for having several args for the same event which originate on different sites, for example). So in other words, we'd set the entity value not just be supplying the ID, but the GUID.

The issue there though is that we then have to take the full GUID into account when retrieving an entity, instead of just the ID as we currently do. We could get around that by requiring the whole entity object to be passed in when setting the context. We could also temporarily switch contexts to the specified one so that then it is current when we grab the entity data from the database (and then switch back to the previous context). I guess that is basically what we'd have to do to retrieve the object to pass in or to get it in our getter anyway. So I guess it makes sense to make it a generic part of the API. (And context switching might have other uses too.)

Another thing to consider is that when we're pulling up related entities we'll need to be able to specify the context that the related entity comes from. Usually entities are related to other entities from the same context, so by default I suppose that contexts should sort of "cascade" down the relationship line. Otherwise when we get an entity from a "foreign" context we'd get the incorrect data when accessing its relationships.

**Comment by JDGrimes** [comment]
## Context vs State

It is easy to confound context and state, but they really aren't the same thing. They just happen to both be mixed together in the most common case that we have to deal with right now (multisite). The context of an entity could be specified on that item's object when it is retrieved from the database, and then it would automatically be stored using the correct context regardless of what the "current" context was (or whether there was a current context set at all. We'll probably be doing something like this with the hook reactions.

In a case like that, we don't need to set the current context to retrieve the data for the entity. In fact, there may be no concept of "current" context at all. So what we said above about always having to switch to a particular context to retrieve the data for an entity isn't strictly true. We do need to switch to another site to retrieve a post from it, but we don't need to switch storage containers to retrieve a different reaction (for example—in case we'd ever want "meta" events).

This leads us to a new dilemma. When there is no concept of "current" context, we can't just get the entity data using the ID. We also need to know the context to pull it from. In other words, the entity really can't be identified by just one value, we need to know the ID and the context to get the entity data. In such a case we must either require the entity object (which includes a context reference) to be passed in, or we must otherwise get passed the GUID somehow.

I guess this sort of has to be left up to each entity class to decide how to determine the GUID when the value is being set. We can default to setting it based on the current context if no context is passed in, but allow each entity to override that behavior if it needs to.

**Comment by JDGrimes** [comment]
## Places GUIDs may be needed

I needed to convince myself that this API was actually going to be useful, so I did a little research. Here are some places that we are using entity IDs but should probably be using the GUIDs:
- Points reactor `hit()` and `spam()` methods.
- Exporting entity data for the UI (enumerable entities).
- Entity array class.

In addition, there are some places that we just get the value of an entityish thing, but should be using the GUID when it is an entity:
- The equals condition.
- The periods extension (I'm not even sure why we're not just using the `get_the_ID()` instead of `get_the_value()` there—are we expecting non-entities can be related to periods?).

**Comment by JDGrimes** [comment]
## Support Non-stateful Contexts?

I think we sort of need to decide whether we need to support non-stateful contexts or not. Supporting them will make things more complex, and I'm not sure there are really any benefits. But I guess that would be similar in many ways to handling out-of-state GUIDs. The only difference would be that we could switch states to get the out-of-state data, whereas we can't do that with non-stateful IDs. I guess the solution is to just allow for both of these possibilities by passing along the entity ID/object to the context retriever. So it could try to determine the context from the entity itself if it needed to, instead of just assuming that the entity itself will never contain any information about the state. Then we'd also need to accept a GUID being passed in, in cases where the entity object doesn't have the context as an attribute, or when we just have the ID. Then, for example, an event arg object could just pass in the GUID instead of trying to retrieve the data for the entity or otherwise setting the context on the entity object for the context object to read (basically a hack).

If we go that direction though, the main takeaway is that we'll be better off encapsulating entity GUIDs into their own objects that can be passed around, instead of trying to guess whether a piece of data is an entity, entity ID, or entity GUID. There are other places that that would be useful other than just when attempting to pass a GUID into `set_the_value()`.

For example, we'd need to set the GUID for each related entity when pulling out the related entity IDs from the database. Otherwise, if they are based on a stateful context, but the context changes before the related entities are actually retrieved, the wrong values will be pulled up. But at that point we don't even know the context to which those entities are related (from within the relationship object, since that data is in the entity objects). So that's a problem. I guess we have to decide how we want the entity API to behave when we're switching contexts. It is hard for us to decide that though, since we don't know the nature of all of the stateful contexts we might be dealing with.

### Conclusion

The only real consensus that I can come to is that it is really too early for us to develop this API very far. Let's go with the bare minimum, and build on that as needed (probably mostly after the initial release, but we can always do more before release if we need to).

So for now we'll just say that contexts are assumed to be stateful, and that behavior when switching contexts is currently undefined and should not be relied upon.

---

Random notes, for posterity when we revisit this some day:
- When pulling the related IDs from the database we need to have the full GUIDs. But can we reliably tell a GUID from a regular ID or array of entities? 
- When a context is nonstateful and the behavior hasn't been properly overridden, we should probably fail if not supplied the context or full GUID when setting the entity value. I guess we'd really have little choice, we couldn't retrieve the entity data.
- We also need to switch contexts so that the context of the primary entity is the current context when we are pulling the related entities from the database.

**Comment by JDGrimes** [comment]
I think we should change the way that we are having entities specify contexts. Since contexts are hierarchical, we are having the entity supply an array of context slugs, which are expected to be in hierarchical order. However, contexts are all registered in the same namespace, so it is somewhat odd that we treat require them to be arranged in a way that seems to imply that they aren't. Also, it is very WET to be supplying the complete list in every location, when we could conceivably just supply the most nested context's slug, and what context that context exists in could be specified in a single location. The `site` context will always have `network` as its parent; every context will always have the same parent. So it would make more sense to have the context object supply the slug of the parent, rather than forcing everybody else in the universe to do that. Also, this would make the API less fragile, since it would conceivably allow us to insert new contexts into the hierarchy in the future, without all of these hard-coded things breaking (though of course every GUID would break, but that's another problem altogether, and assumes that we change this after installing the plugin). I think it is possible that we'd want to add new items to the context, at least that some devs might find this useful in certain situations on custom WordPress installs. An edge case, yes, but the main point is that it is redundant to supply the entire hierarchy everywhere, and could even be a pain to deal with in the future.

**Comment by JDGrimes** [comment]
The main thing that still needs to be done here is to start using the GUIDs where needed. But we'll open new issues for that.

One thing that I'm a little bit unsure about is the fact that there is no main context object. It really isn't needed, I suppose, but it is strange that when specifying the main context we currently use an empty string (`''`). Before when it was an empty array instead it seemed a little more polished. I guess I'll leave it like this for now though. If I can come up with a real reason to change it, I'll open a new ticket.

So I guess that's it. We'll probably revisit context switching at some point in the future, but for now we just need to the first version of this out, and context switching is just a nicety.
